archive/extract_activity.go

```diff
@@ -0,0 +1,177 @@
+	logger := temporal.GetLogger(ctx)
+	logger.V(1).Info("Executing ExtractActivity", "SourcePath", params.SourcePath)
+
+	a.cfg.setDefaults()
+
+	dest, err := a.extract(ctx, params.SourcePath)
+	if err != nil {
+		switch err {
+		case ErrNotAFile:
+			logger.V(2).Info("ExtractActivity: not a file", "SourcePath", params.SourcePath)
+			return nil, err
+		case ErrInvalidArchive:
+			logger.V(2).Info("ExtractActivity: not a valid archive", "SourcePath", params.SourcePath)
+			return nil, err
+		default:
+			return nil, fmt.Errorf("extract archive: %v", err)
+		}
+	}
+
+	dest, err = skipTopLevelDir(dest)
+	if err != nil {
+		return nil, fmt.Errorf("extract archive: skipTopLevelDir: %v", err)
+	}
+
+	return &ExtractActivityResult{DestPath: dest}, nil
+}
+
+func (a *ExtractActivity) extract(ctx context.Context, src string) (string, error) {
+	fi, err := os.Stat(src)
+	if err != nil {
+		return "", err
+	}
+
+	if fi.IsDir() {
+		return src, ErrNotAFile
+	}
+
+	f, err := os.Open(src) // #nosec G304 -- trusted path.
+	if err != nil {
+		return "", fmt.Errorf("open: %v", err)
+	}
+	defer f.Close()
+
+	format, r, err := archiver.Identify(src, f)
+	if err != nil {
+		if errors.Is(err, archiver.ErrNoMatch) {
+			return "", ErrInvalidArchive
+		}
+		return "", fmt.Errorf("identify archive: %v", err)
+	}
+
+	var dest string
+	if ex, ok := format.(archiver.Extractor); ok {
+		dest, err = os.MkdirTemp("", "sdps-extract-archive")
+		if err != nil {
+			return "", fmt.Errorf("create temp dir: %v", err)
+		}
+
+		if err := ex.Extract(ctx, r, nil, a.destFileHandler(dest)); err != nil {
+			return "", fmt.Errorf("extract: %v", err)
+		}
+	} else {
+		return "", fmt.Errorf("no extractor found: %q", src)
+	}
+
+	return dest, nil
+}
+
+func (a *ExtractActivity) destFileHandler(dest string) archiver.FileHandler {
+	return func(ctx context.Context, f archiver.File) error {
+		path := filepath.Join(dest, f.NameInArchive)
+
+		if f.IsDir() {
+			// Make any missing dirs in path, then return.
+			if err := os.MkdirAll(path, a.cfg.DirMode); err != nil {
+				return fmt.Errorf("make directories: %v", err)
+			}
+			return nil
+		} else {
+			// Make any missing parent dirs before creating a file.
+			if err := os.MkdirAll(filepath.Dir(path), a.cfg.DirMode); err != nil {
+				return fmt.Errorf("make parent directories: %v", err)
+			}
+		}
+
+		df, err := safeopen.CreateBeneath(dest, f.NameInArchive)
+		if err != nil {
+			return fmt.Errorf("create file: %v", err)
+		}
+		defer df.Close()
+
+		if err := df.Chmod(a.cfg.FileMode); err != nil {
+			return fmt.Errorf("chmod: %v", err)
+		}
+
+		r, err := f.Open()
+		if err != nil {
+			return fmt.Errorf("open source file: %v", err)
+		}
+		defer r.Close()
+
+		_, err = io.Copy(df, r)
+		if err != nil {
+			return fmt.Errorf("copy: %v", err)
+		}
+
+		return nil
+	}
+}
+
+func skipTopLevelDir(base string) (string, error) {
+	items, err := os.ReadDir(base)
+	if err != nil {
+		return "", fmt.Errorf("read dir: %v", err)
+	}
+
```

Comment:
Should we check for zero items and return an error?

Reply:
Sure. I think the chances of receiving an empty archive is pretty small, but I guess it's possible.

Reply:
On second thought, I don't think we should concern ourselves with an empty archive here.  We are returning the base extract path if there isn't exactly one sub-directory in the extract directory which I think is all this function should handle. Even at the level of the extract activity, I don't think we should concern ourselves with the archive contents.  If we've successfully extracted the archive, let the caller worry about whether it being empty is a problem.